components/TextFields/src/MDCTextInputControllerLegacyDefault.m

```diff
@@ -0,0 +1,1261 @@
+  // to a Y that works. In other words, no bottom inset will make a difference here for UITextFields
+  UIEdgeInsets textInsets = defaultInsets;
+
+  if (!self.isFloatingEnabled) {
+    return defaultInsets;
+  }
+
+  textInsets.top = MDCTextInputDefaultVerticalPadding +
+                   MDCRint(self.textInput.placeholderLabel.font.lineHeight *
+                           (CGFloat)self.floatingPlaceholderScale.floatValue) +
+                   MDCTextInputDefaultVerticalHalfPadding;
+
+  // The amount of space underneath the underline is variable. It could just be
+  // MDCTextInputDefaultVerticalPadding or the biggest estimated underlineLabel height +
+  // MDCTextInputDefaultVerticalHalfPadding
+  CGFloat underlineLabelsOffset = 0;
+  if (self.textInput.leadingUnderlineLabel.text.length) {
+    underlineLabelsOffset =
+        MDCCeil(self.textInput.leadingUnderlineLabel.font.lineHeight * 2.f) / 2.f;
+  }
+  if (self.textInput.trailingUnderlineLabel.text.length || self.characterCountMax) {
+    underlineLabelsOffset =
+        MAX(underlineLabelsOffset,
+            MDCCeil(self.textInput.trailingUnderlineLabel.font.lineHeight * 2.f) / 2.f);
+  }
+  CGFloat underlineOffset = MDCTextInputDefaultVerticalHalfPadding + underlineLabelsOffset;
+
+  // .bottom = underlineOffset + the half padding above the line but below the text field
+  textInsets.bottom = underlineOffset + MDCTextInputDefaultVerticalHalfPadding;
+
+  return textInsets;
+}
+
+#pragma mark - UITextField & UITextView Notification Observation
+
+- (void)textInputDidBeginEditing:(NSNotification *)note {
+  [CATransaction begin];
+  [CATransaction setAnimationDuration:MDCTextInputDefaultFloatingPlaceholderUpAnimationDuration];
+  [CATransaction
+      setAnimationTimingFunction:[CAMediaTimingFunction
+                                     mdc_functionWithType:MDCAnimationTimingFunctionEaseInOut]];
+
+  [self updateLayout];
+
+  if (self.isFloatingEnabled && self.textInput.text.length == 0) {
+    [self movePlaceholderToUp:YES];
+  }
+  [CATransaction commit];
+
+  if (self.characterCountMax > 0) {
+    NSString *announcementString;
+    if (!announcementString.length) {
+      announcementString = [NSString
+          stringWithFormat:@"%lu character limit.", (unsigned long)self.characterCountMax];
+    }
+
+    // Simply sending a layout change notification does not seem to
+    UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, announcementString);
+  }
+}
+
+- (void)textInputDidChange:(NSNotification *)note {
+  if ([note.name isEqualToString:MDCTextFieldTextDidSetTextNotification]) {
+    [CATransaction begin];
+    [CATransaction setAnimationDuration:0];
+    [self updatePlaceholderY];
+    [CATransaction commit];
+  } else {
+    [self updateLayout];
+  }
+
+  // Accessibility
+  if (self.textInput.isEditing && self.characterCountMax > 0) {
+    NSString *announcementString;
+    if (!announcementString.length) {
```

Comment:
announcementString was just created / set to nil on the line above.
Will announcementString.length ever not be 0?
Do we mean to test another string here?